src/modules/payload_deliverer/payload_deliverer.cpp

```diff
@@ -0,0 +1,255 @@
+	return true;
+}
+
+bool PayloadDeliverer::initialize_gripper()
+{
+	// If gripper instance is invalid, try initializing it
+	if (!_gripper.is_valid() && _param_gripper_enable.get()) {
+		GripperConfig config{};
+		config.type = (GripperConfig::GripperType)_param_gripper_type.get();
+		config.sensor = GripperConfig::GripperSensorType::NONE; // Feedback sensor isn't supported for now
+		config.timeout_us = hrt_abstime(_param_gripper_timeout_s.get() * 1000000ULL);
+		_gripper.init(config);
+	}
+
+	// NOTE: Support for changing gripper sensor type / gripper type configuration when
+	// the parameter change is detected isn't added as we don't have actual use case for that
+	// yet!
+
+	if (!_gripper.is_valid()) {
+		PX4_DEBUG("Gripper object initialization invalid!");
+		return false;
+
+	} else {
+		_gripper.grab(); // Initialize to grab position
+		return true;
+	}
+}
+
+void PayloadDeliverer::parameter_update()
+{
+	// Call the ModuleParams function to update internal parameters
+	updateParams();
+	initialize_gripper();
+}
+
+void PayloadDeliverer::Run()
+{
+	if (should_exit()) {
+		ScheduleClear();
+		exit_and_cleanup();
+		return;
+	}
+
+	vehicle_command_s vcmd{};
+	hrt_abstime now = hrt_absolute_time();
+
+	if (_vehicle_command_sub.update(&vcmd)) {
+		update_gripper(now, &vcmd);
+
+	} else {
+		update_gripper(now);
+
+	}
+}
+
+void PayloadDeliverer::update_gripper(const hrt_abstime &now,  const vehicle_command_s *vehicle_command)
+{
+	if (!_param_gripper_enable.get()) {
+		PX4_WARN("Gripper isn't enabled, so the gripper functionality will not work!");
+		// If gripper isn't enabled, don't do anything
+		return;
+	}
+
+	_gripper.update();
+
+	// Process successful release command acknowledgement
+	if (_gripper.released_read_once()) {
+		vehicle_command_ack_s vcmd_ack{};
+		vcmd_ack.timestamp = now;
+
+		vcmd_ack.command = vehicle_command_s::VEHICLE_CMD_DO_GRIPPER;
+		vcmd_ack.result = vehicle_command_ack_s::VEHICLE_CMD_RESULT_ACCEPTED;
+		_vehicle_command_ack_pub.publish(vcmd_ack);
+
+		PX4_DEBUG("Payload Drop Successful Ack Sent!");
+	}
+
+	// If there's no vehicle command to process, just return
+	if (vehicle_command == nullptr) {
+		return;
+	}
+
+	// Process if we received DO_GRIPPER vehicle command
+	if (vehicle_command -> command == vehicle_command_s::VEHICLE_CMD_DO_GRIPPER) {
+		PX4_DEBUG("Gripper command received!");
+		const int32_t gripper_action = *(int32_t *)&vehicle_command -> param2; // Convert the action to integer
+
+		switch (gripper_action) {
+		case vehicle_command_s::GRIPPER_ACTION_GRAB:
+			_gripper.grab();
+			break;
+
+		case vehicle_command_s::GRIPPER_ACTION_RELEASE:
+			_gripper.release();
+			break;
+		}
+	}
+}
+
+void PayloadDeliverer::gripper_test()
+{
+	if (!_gripper.is_valid()) {
+		PX4_DEBUG("Gripper is not initialized correctly!");
+		return;
+	}
+
+	PX4_DEBUG("Test: Opening the Gripper!");
+	_gripper.release();
+	px4_usleep(5_s);
+	PX4_DEBUG("Test: Closing the Gripper!");
+	_gripper.grab();
```

Comment:
Better publish a vehicle command

Reply:
But that wouldn't work since the `gripper_test` function would be blocking (the `Run()` execution would come to halt) while it is in use, and hence the vehicle command won't get registered and actuate the gripper (which is done in `Run()`).

I have just tested this by activating the command, and I can verify that actuator doesn't move (due to the reason given above, I assume).

Or am I missing something?

Reply:
![image](https://user-images.githubusercontent.com/23277211/183675407-b2febff6-d5ab-4352-a092-f2d6a9387d82.png)

As shown here, using the vehicle command, it doesn't get translated into the gripper command as expected while executing the gripper_test.

So I think sticking to utilizing `_gripper` object for grab/releasing is ok, what do you think?

Reply:
`custom_command` is executed on a different thread so it should work.

Reply:
I will try again, but it seemed like the program was on hold while the gripper_test() function was executing 🤔

Reply:
Solved the problem in the latest commit here: https://github.com/PX4/PX4-Autopilot/pull/19963/commits/9e73973dc4ccc982df9ae408e4f4d91dd7a5a541

There was a bug on my side, due to the SubscriptionCallback usage. Details are explained in the commit.